Mathlib/RingTheory/Presentation.lean

```diff
@@ -0,0 +1,220 @@
+  finite_relations : Finite P.relations
+
+attribute [instance] IsFinite.finite_vars IsFinite.finite_relations
+
+lemma ideal_fg_of_isFinite [P.IsFinite] : P.ker.FG := by
+  use (Set.finite_range P.relation).toFinset
+  simp [ideal_eq_span_range_relation]
+
+/-- If a presentation is finite, the corresponding quotient is
+of finite presentation. -/
+instance [P.IsFinite] : FinitePresentation R P.Quotient :=
+  FinitePresentation.quotient P.ideal_fg_of_isFinite
+
+lemma finitePresentation_of_presentation_of_isFinite [P.IsFinite] :
+    FinitePresentation R S :=
+  FinitePresentation.equiv (P.quotientEquiv.restrictScalars R)
+
+section Construction
+
+section Localization
+
+variable (r : R) [IsLocalization.Away r S]
+
+open IsLocalization.Away
+
+private lemma ker_algebraMap_eq_span_range_relation_localizationAway :
+    RingHom.ker (aeval (S₁ := S) (Generators.localizationAway r).val) =
+      Ideal.span { C r * X () - 1 } := by
+  have : aeval (S₁ := S) (Generators.localizationAway r).val =
+      (mvPolynomialQuotientEquiv S r).toAlgHom.comp
+        (Ideal.Quotient.mkₐ R (Ideal.span {C r * X () - 1})) := by
+    ext x
+    simp only [Generators.localizationAway_vars, aeval_X, Generators.localizationAway_val,
+      AlgEquiv.toAlgHom_eq_coe, AlgHom.coe_comp, AlgHom.coe_coe, Ideal.Quotient.mkₐ_eq_mk,
+      Function.comp_apply]
+    rw [IsLocalization.Away.mvPolynomialQuotientEquiv_apply, aeval_X]
+  rw [this]
+  erw [← RingHom.comap_ker]
+  simp only [Generators.localizationAway_vars, AlgEquiv.toAlgHom_eq_coe, AlgHom.toRingHom_eq_coe,
+    AlgEquiv.toAlgHom_toRingHom]
+  show Ideal.comap _ (RingHom.ker (mvPolynomialQuotientEquiv S r)) =
+    Ideal.span {C r * X () - 1}
+  simp [RingHom.ker_equiv, ← RingHom.ker_eq_comap_bot]
+
+/-- If `S` is the localization of `R` away from `r`, we can construct a natural
+presentation of `S` as `R`-algebra with a single generator `X` and the relation `r * X - 1 = 0`. -/
+@[simps relations relation]
+noncomputable def localizationAway : Presentation R S where
+  toGenerators := Generators.localizationAway r
+  relations := Unit
+  relation _ := C r * X () - 1
+  ker_algebraMap_eq_span_range_relation := by
+    simp only [Generators.localizationAway_vars, Set.range_const]
+    apply ker_algebraMap_eq_span_range_relation_localizationAway r
+
+instance localizationAway_isFinite : (localizationAway r (S := S)).IsFinite where
+  finite_vars := inferInstanceAs <| Finite Unit
+  finite_relations := inferInstanceAs <| Finite Unit
+
+@[simp]
+lemma localizationAway_dimension_zero : (localizationAway r (S := S)).dimension = 0 := by
+  simp [Presentation.dimension, localizationAway]
+
+end Localization
+
+variable {T} [CommRing T] [Algebra R T] (P : Presentation R S)
+
+private lemma ker_algebraMap_eq_span_range_relation_baseChange :
+    RingHom.ker (aeval (R := T) (S₁ := T ⊗[R] S) P.baseChange.val) =
+      Ideal.span (Set.range fun i ↦ (MvPolynomial.map (algebraMap R T)) (P.relation i)) := by
```

Comment:
One could write the `fun i ↦ ...` part more concisely as `(MvPolynomial.map (algebraMap R T)) ∘ P.relation`, but I haven't check if this makes things better or worse later.

Reply:
It seems to make things worse, so I left it as is.

Reply:
An alternative (which is just alternative notation, AFAIK) is
`((MvPolynomial.map (algebraMap R T)) (P.relation ·))` .

Reply:
(But this is not really important...)

Reply:
The scope of `·` seems to be interpreted as restricted to `P.relation`, so this causes a type error.